middlewares/sticky_backend_handler.go

```diff
@@ -0,0 +1,274 @@
+package middlewares
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"html/template"
+	"io"
+	"net"
+	"net/http"
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/sprig"
+	"github.com/containous/traefik/healthcheck"
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/types"
+	guuid "github.com/satori/go.uuid"
+	"github.com/serialx/hashring"
+	"github.com/vulcand/oxy/forward"
+)
+
+// StickinessParsed holds the parsed representation of a Stickiness object
+type StickinessParsed struct {
+	BackendName      string
+	CookieEncryptKey string
+	CookieName       string
+	Rules            []*template.Template
+	UseCookie        bool
+	UseIP            bool
+	UseRules         bool
+}
+
+// NewStickinessParsed creates a new StickinessParsed instance
+func NewStickinessParsed(s *types.Stickiness, backendName string, cookieName string) *StickinessParsed {
+	var useCookie, useIP, useRules bool
+	rules := make([]*template.Template, 0)
+	useDefault := true
+	if len(s.Rules) > 0 {
+		useRules = true
+		useDefault = false
+		for i, rule := range s.Rules {
+			t, err := template.New(guuid.NewV4().String()).Funcs(sprig.FuncMap()).Parse(rule)
+			if err != nil {
+				log.Errorf("Backend %s: failed to parse sticky rule with index %d, error: %s", backendName, i, err)
+				t = nil
+			}
+			rules = append(rules, t)
+		}
+	}
+	if s.IP {
+		useIP = true
+		useDefault = false
+	}
+	if s.Cookie || useDefault {
+		useCookie = true
+	}
+
+	return &StickinessParsed{
+		BackendName:      backendName,
+		CookieEncryptKey: s.CookieEncryptKey,
+		CookieName:       cookieName,
+		Rules:            rules,
+		UseCookie:        useCookie,
+		UseIP:            useIP,
+		UseRules:         useRules,
+	}
+}
+
+// StickyBackendHandler inspects criteria such as Client IP address and uses
+// consistent hashing to route a request to a specific backend server
+type StickyBackendHandler struct {
+	lb   healthcheck.LoadBalancer
+	next http.Handler
+	sp   *StickinessParsed
```

Comment:
Rename `sp` to `config` to improve readability (relates to previous comment about `StickinessParsed` being renamed to `StickinessConfig`